File: matlab/demos/jmm/demo_jmm_3_3_xcorr.m
%% demo_jmm_3_3_xcorr.m
% Analysis 3.3 (Online Supplement, Section 10): phase as lag, a windowed
% cross-correlogram.
%
% A demo of the Music Perception Toolbox reproducing the analysis from the
% JMM article; lightly edited from the article's own script. Data come
% from the jmm package (BWV 347 read from the bundled MusicXML) or
% jmm.pianoPhase (the rendered Piano Phase voices); the figures stay on screen unless
% SAVE_FIGURES is set.
%
% Analysis 3.3: phase as lag in Reich's Piano Phase.
%
% The query is a single canonical cell (Piano 1's repeating pattern); the
% context is Piano 2's event stream. At each anchor (a Piano-1 cell
% boundary) the query is translated in time by a lag tau spanning one
% whole cell, and the one-sided matched-filter response against Piano 2
% is read off (simMaet with 'normalize', 'oneSidedDenom', the
% Analysis-1.4 idiom). Collecting these rows gives a cross-correlogram
% R(anchor, tau); its bright ridge tracks the running phase offset
% between the two pianos, climbing the staircase 0 -> 12 pulses across
% the piece. Secondary ridges a half-cell away are expected from the
% cell's near-period-6 internal self-similarity.
%
% Pre-MAET structure:
%
%     attribute   sigma           rel   per
%     ---------   --------------  ----  ----
%     pitch       0.15 semitone   no    no
%     time        0.015 s         no    no
%
%     r = (1, 1); query = 12-event canonical cell; context = Piano 2;
%     one-sided normalization (divide by the query self-overlap).
%
% Data: jmm.pianoPhase (the rendered Piano Phase voices). Toolbox:
% preMaetFromAttrTable, packPreMaet, windowedSimilarity. Runtime: half a
% minute.

% The demo folder is located from the toolbox root, and adding it puts
% the +jmm helper package in scope.
mptRoot = which('buildMaet');
if isempty(mptRoot)
    error('demoJmm:toolboxNotFound', ...
        ['The toolbox is not on the path. Add the matlab folder of the ' ...
         'Music Perception Toolbox, then run this demo again.']);
end
thisDir = fullfile(fileparts(mptRoot), 'demos', 'jmm');
addpath(thisDir);
clear mptRoot

% Set true to write the figures to a
% figures/ folder beside this script; false leaves them on screen only.
SAVE_FIGURES = false;

prevDefaults = mptDefaults('showHints', false, 'truncationSigmas', 4.0);

% --- parameters --------------------------------------------------------------
SIGMA_PITCH = 0.15;
SIGMA_TIME  = 0.015;                % 15 ms
pe = jmm.pianoPhase();
IOI         = pe.baseIoi;
CELL_DUR    = pe.nc * IOI;          % one cell in seconds
N_TAU       = 241;                  % lag samples over one cell (resolves ~33 ms ridge)
ASTEP       = 2;                    % anchor every ASTEP Piano-1 cells

isRel   = [false false];
isPer   = [false false];
periods = [0 0];
sigma   = [SIGMA_PITCH, SIGMA_TIME];
rVec    = [1 1];

% --- query: one canonical cell (Piano 1's pattern), cell starting at t = 0 --
qPitch = double(pe.cell);                       % 1 x 12
qTime  = (0:(pe.nc - 1)) * IOI;                 % 1 x 12
queryPAttr = {qPitch, qTime};

% --- context: Piano 2 ---------------------------------------------------------
context = preMaetFromAttrTable(pe.voice2Table, 'attributes', { ...
    struct('column', 'pitch', 'sigma', sigma(1)), ...
    struct('column', 'onset', 'name', 'time', 'sigma', sigma(2))}, ...
    'time', 'seconds', 'chords', 'separate', 'weights', 'ones');
ctxPAttr = unpackPreMaet(context);
t2 = ctxPAttr{2};

% --- anchors and lag grid -----------------------------------------------------
nCells = pe.nRepsV1;
anchors = (0:ASTEP:(nCells - 1)) * CELL_DUR;
tauGrid = (0:(N_TAU - 1)) * (CELL_DUR / N_TAU);  % linspace without the endpoint

% Single windowedSimilarity call producing the whole (anchor, lag)
% correlogram. The context (Piano 2) is localized by a rectangular window
% of full support 2 * HALF at each anchor; the query is translated to each
% lag. The output shape follows queryCentres: an (anchors x N_TAU) matrix
% of absolute query centres gives an (anchors x N_TAU) response, with the
% context broadcast along the lag axis.
%
% Query placement. windowedSimilarity translates the query so its mean on
% the time axis lands at queryCentres(i, j); the original loop translated
% the query by (a - tau) directly, i.e. landed its mean at (a - tau) + mu_q.
% So queryCentres = (a - tau) + mu_q reproduces the original placement.
% Piano 2 leads, so retarding the query by tau makes the ridge read the
% running phase k directly.
HALF = CELL_DUR + 2 * IOI;
muQ = mean(qTime);
queryCentres = (anchors(:) - tauGrid) + muQ;    % (anchors, N_TAU)
% The query is read under the context's geometry, so it carries the same
% specs.
[~, ~, ctxSpecs] = unpackPreMaet(context);
query = packPreMaet(queryPAttr, [], ctxSpecs);
showPreMaet(context, 'maxEvents', 4);
showPreMaet(query, 'maxEvents', 4);

R = windowedSimilarity( ...
    context, query, anchors, ...
    'queryCentres', queryCentres, ...
    'contextWindow', {1.0, 2 * HALF}, ...       % rectangle, full support 2 * HALF
    'normalize', 'oneSidedDenom', ...
    'windowAttr', 2, ...
    'dropWindowAttr', false, ...
    'verbose', false);
% Reproduce the original sparse-context skip: blank anchors whose
% localization window holds fewer than one full cell of context events.
ctxCounts = arrayfun(@(a) sum((t2 >= a - HALF) & (t2 <= a + HALF)), anchors);
R(ctxCounts < pe.nc, :) = NaN;

% --- true lag staircase for overlay ------------------------------------------
trueLag = pe.lagAt(anchors / CELL_DUR);

% --- figure ------------------------------------------------------------------
fig = figure('Position', [50 50 1300 460], 'Color', 'w');
ax = axes('Parent', fig, 'Position', [0.06 0.15 0.80 0.75]);
imagesc(ax, [anchors(1), anchors(end)], [0, pe.nc], R.');
set(ax, 'YDir', 'normal');
colormap(ax, jmm.colourMap('magma'));
caxis(ax, [0, max(R(:))]);
hold(ax, 'on');
hLag = plot(ax, anchors, trueLag, 'Color', [0.224 0.827 1.0], 'LineWidth', 1.1);
for s = pe.shiftCentres
    plot(ax, [s s], [0, pe.nc], 'Color', [0.55 0.55 0.55], 'LineWidth', 0.4);
end
xlabel(ax, 'anchor time \alpha (s)', 'FontSize', 15);
ylabel(ax, 'lag \tau (pulses)', 'FontSize', 15);
set(ax, 'YTick', 0:2:pe.nc, 'FontSize', 13);
title(ax, ['Lag cross-correlation: canonical cell (Piano 1) vs Piano 2 ' ...
           '(one-sided matched filter)'], 'FontSize', 16);
legend(hLag, {'true phase k'}, 'Location', 'northwest', 'FontSize', 12);
cb = colorbar(ax);
ylabel(cb, 'matched-filter response', 'FontSize', 13);

figDir = fullfile(thisDir, 'figures');
if SAVE_FIGURES && ~exist(figDir, 'dir'), mkdir(figDir); end
if SAVE_FIGURES
    print(fig, '-dpng', '-r140', fullfile(figDir, 'demo_jmm_3_3_xcorr.png'));
    fprintf('Saved figures/demo_jmm_3_3_xcorr.png\n');
end
Rf = R(~isnan(R));
fprintf('R range %.3f-%.3f, %d anchors x %d lags\n', ...
        min(Rf), max(Rf), numel(anchors), N_TAU);

% The demo leaves the toolbox as it found it: the defaults it set at the
% top are restored here.
mptDefaults(prevDefaults);

File: python/demos/jmm/demo_jmm_3_3_xcorr.py
"""demo_jmm_3_3_xcorr.py — Analysis 3.3 (Section 4.3.3 of the JMM article).

A demo of the Music Perception Toolbox reproducing the analysis from the
JMM article; lightly edited from the article's own script. Data come
from jmm_data (BWV 347 read from the bundled MusicXML) or piano_phase
(the rendered Piano Phase voices); a figure is written to figures/ when
matplotlib is available.

Analysis 3.3: phase as lag in Reich's *Piano Phase*.

The query is a single canonical cell (Piano 1's repeating pattern); the context
is Piano 2's event stream. At each anchor (a Piano-1 cell boundary) the query is
translated in time by a lag tau spanning one whole cell, and the one-sided
matched-filter response against Piano 2 is read off (cos_sim_exp_tens with
normalize='oneSidedDenom', the Analysis-1.4 idiom). Collecting these rows gives
a cross-correlogram R(anchor, tau); its bright ridge tracks the running phase
offset between the two pianos, climbing the staircase 0 -> 12 pulses across the
piece. Secondary ridges a half-cell away are expected from the cell's
near-period-6 internal self-similarity.

Pre-MAET structure::

    attribute   sigma           rel   per
    ---------   --------------  ----  ----
    pitch       0.15 semitone   no    no
    time        0.015 s         no    no

    r = (1, 1); query = 12-event canonical cell; context = Piano 2;
    one-sided normalisation (divide by the query self-overlap).
"""

import numpy as np
try:
    import matplotlib.pyplot as plt
except ImportError:
    plt = None
plt.rcParams.update({'font.size': 15, 'axes.titlesize': 17, 'axes.labelsize': 15,
                     'xtick.labelsize': 13, 'ytick.labelsize': 13, 'figure.titlesize': 20,
                     'font.family': 'DejaVu Sans'})

import mpt
mpt.set_default(show_hints=False, truncation_sigmas=4.0, kernel_precision='double')
from mpt import show_pre_maet, windowed_similarity

import piano_phase as pe

# --- parameters ------------------------------------------------------------
SIGMA_PITCH = 0.15
SIGMA_TIME  = 0.015                 # 15 ms
IOI         = pe.BASE_IOI
CELL_DUR    = pe.NC * IOI           # one cell in seconds
N_TAU       = 241                   # lag samples over one cell (resolves ~33 ms ridge)
ASTEP       = 2                     # anchor every ASTEP Piano-1 cells

is_rel   = [False, False]
is_per   = [False, False]
periods  = [0.0, 0.0]
sigma    = [SIGMA_PITCH, SIGMA_TIME]
r_vec    = [1, 1]

# --- query: one canonical cell (Piano 1's pattern), cell starting at t = 0 --
q_pitch = pe.CELL.astype(float).reshape(1, pe.NC)
q_time  = (np.arange(pe.NC) * IOI).reshape(1, pe.NC)
query_pattr = [q_pitch, q_time]

# --- context: Piano 2 -------------------------------------------------------
p2, t2 = pe.render_voice(2)

# --- anchors and lag grid ---------------------------------------------------
n_cells = pe.N_REPS_V1
anchors = np.arange(0, n_cells, ASTEP) * CELL_DUR
tau_grid = np.linspace(0.0, CELL_DUR, N_TAU, endpoint=False)

# Single windowed_similarity call producing the whole (anchor, lag)
# correlogram. The context (Piano 2) is localised by a rectangular window
# of full support 2 * HALF at each anchor; the query is translated to each
# lag. The output shape follows query_centres: an (anchors x N_TAU) matrix
# of absolute query centres gives an (anchors x N_TAU) response, with the
# context broadcast along the lag axis.
#
# Query placement. windowed_similarity translates the query so its mean on
# the time axis lands at query_centres[i, j]; the original loop translated
# the query by (a - tau) directly, i.e. landed its mean at (a - tau) + mu_q.
# So query_centres = (a - tau) + mu_q reproduces the original placement.
# Piano 2 leads, so retarding the query by tau makes the ridge read the
# running phase k directly.
HALF = CELL_DUR + 2 * IOI
mu_q = float(query_pattr[1].mean())
query_centres = (anchors[:, None] - tau_grid[None, :]) + mu_q   # (anchors, N_TAU)
show_pre_maet([p2.reshape(1, -1), t2.reshape(1, -1)], None,
              names=['pitch', 'onset'], sigma=sigma, is_rel=is_rel,
              is_per=is_per, period=periods, max_events=4)
show_pre_maet(query_pattr, None, names=['pitch', 'onset'], sigma=sigma,
              is_rel=is_rel, is_per=is_per, period=periods, max_events=4)

R = windowed_similarity(
    [p2.reshape(1, -1), t2.reshape(1, -1)], None,
    query_pattr, None,
    sigma, r_vec, is_rel, is_per, periods,
    anchors,
    query_centres=query_centres,
    context_window=(1.0, 2 * HALF),     # rectangle, full support 2 * HALF
    normalize='oneSidedDenom',
    window_attr=1,
    drop_window_attr=False,
    verbose=False,
)
# Reproduce the original sparse-context skip: blank anchors whose
# localisation window holds fewer than one full cell of context events.
ctx_counts = np.array([((t2 >= a - HALF) & (t2 <= a + HALF)).sum()
                       for a in anchors])
R[ctx_counts < pe.NC, :] = np.nan

# --- true lag staircase for overlay ----------------------------------------
true_lag = np.array([pe.lag_at(a / CELL_DUR) for a in anchors])

# --- figure ----------------------------------------------------------------
if plt is None:
    print('matplotlib not available; skipping the figure.')
    raise SystemExit(0)

fig, ax = plt.subplots(figsize=(13, 4.6))
extent = [anchors[0], anchors[-1], 0, pe.NC]
im = ax.imshow(R.T, origin='lower', aspect='auto', extent=extent,
               cmap='magma', vmin=0.0)
ax.plot(anchors, true_lag, color='#39d3ff', lw=1.1, alpha=0.8,
        label='true phase $k$')
for s in pe.shift_centre_times():
    ax.axvline(s, color='w', lw=0.4, alpha=0.3)
ax.set_xlabel(r'anchor time $\alpha$ (s)')
ax.set_ylabel('lag $\\tau$ (pulses)')
ax.set_yticks(range(0, pe.NC + 1, 2))
ax.set_title('Lag cross-correlation: canonical cell (Piano 1) vs Piano 2 '
             '(one-sided matched filter)')
ax.legend(loc='upper left', fontsize=12, framealpha=0.6)
fig.colorbar(im, ax=ax, label='matched-filter response', pad=0.06)
fig.tight_layout()
fig.savefig('figures/demo_jmm_3_3_xcorr.png', dpi=140, bbox_inches='tight')
print(f'saved; R range {np.nanmin(R):.3f}-{np.nanmax(R):.3f}, '
      f'{len(anchors)} anchors x {N_TAU} lags')
